### ISSUE TITLE
pytest session timeout support. 

### ISSUE THREAD

**Issue by anonyknight** [pytest-dev/pytest-timeout#60]
The plugin timeouts a test right now. Is there any reason why it did not support to support session?

Like @pytest.fixture(scope="session")

We can use this statement to stop the whole session.
`request.session.shouldstop = 'time limit reached: %0.2f seconds' % elapsed`

The previous discussion was here:
https://mail.python.org/pipermail/pytest-dev/2016-April/003564.html

Using CI to timeout pytest cannot do a graceful clean and test result report. Thank you.

**Comment by flub** [comment]
> We can use this statement to stop the whole session.
> `request.session.shouldstop = 'time limit reached: %0.2f seconds' % elapsed`

Could you describe a bit more the entire process you mean here?  Not sure I follow.

> Using CI to timeout pytest cannot do a graceful clean and test result report. Thank you.

Have you considered looking if it's possible to get the test reports written out on external termination instead?

**Comment by anonyknight** [comment]
Hi @flub , thanks for your reply and help.

I tried to replace pytest.fail with request.session.shouldstop in the following function to do the concept proof that pytest-timeout could work for pytest session timeout:
```Python
def timeout_sigalrm(item, timeout):
    """Dump stack of threads and raise an exception
    This will output the stacks of any threads other then the
    current to stderr and then raise an AssertionError, thus
    terminating the test.
    """
    if SUPPRESS_TIMEOUT:
        return
    __tracebackhide__ = True
    nthreads = len(threading.enumerate())
    if nthreads > 1:
        write_title("Timeout", sep="+")
    dump_stacks()
    if nthreads > 1:
        write_title("Timeout", sep="+")
    pytest.fail("Timeout >%ss" % timeout)
```
**The timeout mechanism can stop the whole pytest session.**
**But I was not sure the original design why pytest-timeout only timeout a test but not providing another option to timeout the whole session.**

**The use case here:**
CI(Jenkins) set a hard timeout, say 5 hrs for pytest jobs. pytest with --junit-xml reports the whole results only after test suite. If the CI timeout was hit and the whole pytest was killed by an external signal, junit test reports were lost. 

**Two fixes to consider:**
1. Have a pytest session timeout mechanism and gracefully cleanup for a test suite. Avoid using CI to set external timeout and kill pytest process.
2. junitxml plugin report test result after each test, not the whole test suite.

The first fix could fit pytest-timeout domain. That's why I posted an issue here. 

The second fix could be in junitxml.py in pytest.

Thank you very much for your reading and time.
